text/0000-sbom.md

```diff
@@ -0,0 +1,118 @@
+> We propose that increased supply chain transparency through SBOMs can reduce cybersecurity risks and overall costs by:
+> - Enhancing the identification of vulnerable systems and the root cause of incidents
+> - Reducing unplanned and unproductive work
+> - Supporting more informed market differentiation and component selection
+> - Reducing duplication of effort by standardizing formats across multiple sectors
+> - Identifying suspicious or counterfeit software components
+
+Currently since container images built by Cloud Native Buildpacks don't follow a generic file system layout, they may not be easily scannable by container scanning tools. We do however provide a BOM which should greatly help sidestep the entire manual scanning process and speed up things like CVE detection by directly providing the CVE scanner with a BOM.
+
+However, since we currently do not impose any standard for what the BOM should look like and since the metadata table in BOM is a freeform table, it is very hard to have consistent BOMs that can be used by CVE scanners.
+
+Having a structured and consistent SBOM that can be consumed by downstream systems could greatly enhance the security proposition around buildpacks.
+
+## What use cases does it support?
+
+- Consistent SBOM across buildpacks
+- Improved integration with other supply chain analysis and scanning tools
+
+## What is the expected outcome?
+
+- Use of CycloneDX as the SBOM storage format and updates the the specification and lifecycle to reflect the same.
+- Moving the SBOM from a label to a file in a separate layer.
+- Enhancements to the platform spec and `pack` to allow easy interaction with SBOMs including conversion between major SBOM formats.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Currently the SBOM is defined in the `<layer>.toml`, `launch.toml` and `build.toml` respectively under the `[bom]` table.
+It may not be the most user-friendly way for buildpack authors to create SBOM documents in the above format. This RFC proposes that the `bom` be moved to `<layer>.bom.<ext>`, `launch.bom.<ext>` and `build.bom.<ext>` instead where `<ext>` may be `xml` or `json`. This is mostly to allow for different BOM generation tools to be used. This also allows us in the future to add support for `spdx`  tag documents or other file formats if needed.
+
+The above bom documents need to be created in CycloneDX format.
+
+The lifecycle of the respective `bom` files during rebuilds would be exactly the same as the `bom` table currently. 
+
+The lifecycle would be responsible for taking all of the above `bom` files, merging it and storing it in `/layers/config/sbom.json`. This file would also be stored in a layer of its own and the lifecycle would be responsible for storing the layer digest in a label `io.buildpacks.bom.digest`.
```

Comment:
Merging the files makes me worry a bit (maybe unnecessarily?). I start thinking about potential edge cases. Maybe it's unnecessary. What if each layer just had its own individual file. Then no merging is required. Instead of a single file like `sbom.json`, we could have `/layers/config/sbom/<layer>.json`.

I guess the downside of this would be that scanners would need to process multiple files, but that seems like something reasonable. It also seems like it would provide more detail, because the scanner could then know specifically which layer (and could trace back to the buildpack) that installed a particular thing. Maybe they could figure that out another way, but this would make it obvious.

Reply:
I don't know enough about this tooling. So I might be missing the mark here.

Layer level BOMs could be beneficial for automated analysis of each layer.

You can also create a nested BOM. With multiple layers as components under the top level container image.

And yet another option is to create separate BOMs for each layer. And then one final BOM that references all the individual layer BOMs.

They all have different pros and cons. But as someone using a particular container I probably don't care as much about the individual layers. I want to know what I'm using. And I most likely want it as a single BOM.

As someone building a container I probably want to know more per layer detail.

So a combination of the two could be beneficial depending on the use cases you are trying to serve.

Reply:
That's interesting and very flexible. I think my primary concern would be for the support of various tools that will consume this. I see it less as humans reading it and much more as automated tools procured by humans reading it.

Are all of these different ways to represent the layers in the  BOM required to be supported by the CycloneDX specification? Or are some optional under the spec? My worry would be picking a way to represent it, then having popular tool XYZ not support that part of the spec so be unable to read the BOM correctly. Thanks.

Reply:
Yeah, most of the time it should be automated tools.

We don't put restrictions on implementations. It's an open standard that anyone can contribute to, adopt and use. I don't think it's reasonable for us to say you must do X, Y and Z if you are implementing CycloneDX.

But in terms of the simplest, most well supported, approach. I would say a flat list of all components in the final container image. This is the approach that Anchore Syft has taken too.

But as a builder, having the dependency graph information as well is very helpful. It helps you track down where exactly that vulnerable component was pulled in from. And if a tool doesn't support dependency graph information that doesn't matter, it's a separate element to the components and is ignored.

Reply:
This is what the lifecycle would be currently responsible for. After the RFC at https://github.com/buildpacks/rfcs/pull/151

the idea was for the lifecycle to automatically combine the BOM while injecting metadata indicating which layer it came from.  This was meant to solve issues like https://github.com/paketo-buildpacks/packit/issues/54 automatically.

IMHO it makes more sense for a container to have a single SBOM which can have layers as its components or can be a flat list. I would assume most container scanning tools currently produce/consume a single SBOM document and dividing it up into layers might make integrations harder.  Keeping it into a single document while still preserving metadata around which layer contributed to the bom entry might be the best of both worlds imho.

Reply:
From my understanding of buildpacks, each layer represents the output of each buildpack. The reason why I advocate for layer level SBOMs is because it helps with SBOM reuse on the tool side. However, the final output of kpack can combine all the SBOMs into a single one.